### ISSUE TITLE
Intruder energy grain

### ISSUE THREAD

**Issue by Krazyfuq** [ScreepsOCS/screeps.behaviour-action-pattern#478]
When I logged in today one of my LC6 rooms had no creeps and all extensions were empty. Took awhile to figure out why.

A guy was attacking and killing my reserver in another room, this of course was spawning a ranger. Once the reserver was dead the intruder retreated. Due to no more threat the ranger would turn back and suicide at the spawn. Is this intended?
I assume this guy is watching and once a new reserve is spawned and at the controller he sends the attacker to kill it again, so this room has been spawning reservers and rangers who keep suiciding until the room stopped spawning anything.
I have since placed yellow/yellow flag but no defenders have been spawned for this flag, only rangers spawning due to intruder alert. Has yellow/yellow changed, used to work fine.
IMO The rangers spawned for intruders should go to the room of the intruder and defend that room till death. I think this guys is doing this because he knows it is draining my energy.

**Comment by karlthepagan** [comment]
Would be a nice use of the task system: take all military units not handling threats and divert them.

Two things will resolve this : the above relocation and pacifist creeps retreating.

This is however an example of screeps endgame. If you are facing an intelligent enemy you must change your own code to counter patterns they notice in your defense.

**Comment by cyberblast** [comment]
> Due to no more threat the ranger would turn back and suicide at the spawn. Is this intended? 

No, it shouldn't suicide but recycle itself to get your energy back. 

> I have since placed yellow/yellow flag but no defenders have been spawned for this flag 

hmm that's strange, should still work.

> IMO The rangers spawned for intruders should go to the room of the intruder and defend that room till death 

uhm yes, but commonly its better to get that energy back instead letting them idle around once there is no threat left anymore.

**Comment by Dasku** [comment]
I've been thinking about this myself. If there is a genuine threat, what is best, spawning more haulers and miners or just defenders until your room dies? I think defenders is the better option. If you kill the threat your room will survive. If you don't, the invader will kill everything anyway. If its a NPC it will despawn and you can send a pioneer, if its a player, your room is gone. It's fight and win, or die, is always better than surrender.

**Comment by Krazyfuq** [comment]
My point is the defenders suiciding or recycling whatever it is without doing anything, without even reaching the room of the invader. Invader leaves defender suicides, invader returns new defender spawns... rinse and repeat until room can not spawn anything. I'm not a programmer just pointing it out, watching it made no sense to me. The room came to a stop without a battle and could easily have been overtaken, which I don't care it is part of the game and I realize that my non programming skills are what will be my downfall.

The yellow/yellow flag did work but not until the room got back up to a point. I thought all rooms supported the defender flags and was expecting defenders from the other rooms. I eventually pushed him back with the defender flags he countered with dedicated healers for his attackers which I have no counter for that I can figure out, when I try to spawn support from my other rooms, I get null

Game.spawns['Spawn1'].createCreepBySetup(Creep.setup.ranger);
[9:53:55 AM]E71S22 > Hostile intruder ({"tough":1,"move":4,"heal":3}) from "GiantDwarf.
 null

**Comment by cyberblast** [comment]
Yes, sorry about that. We are still working on a better "room selection" where to spawn those defenders (or whatever for any task). It's currently simply the nearest, which is one of the nearest by random (but always the same I guess). 

The problem with those intruders appearing and disappearing is: you can see and follow where they are moving. But the creeps can't. So once the defender spawns, if the intruder can be spotted somewhere, the defender will go there. Even if you don't have sight of that room where it has been spotted initially, the defender should at least move there to have a look if it is still there. 
If there is no intruder in that room, an no intruder is known to be in an adjacent room, there is nothing the defender can do currently. So it will go back home to recycle it back into it's worth in energy. 

If you've got a constant problem in a room, you should put a guard. I agree, that guard may not work quite well, if the next room is a low RCL or nearly dead. As mentioned, we are working on that problem. 

If you want to enforce spawning of a creep, you should use the queue. Using the setup approach (createCreepBySetup), there are a lot of validations which commonly prevent spawning on demand. 
Usage of the queue:
`Game.rooms['<roomName>'].spawnQueueLow.push({parts:[MOVE,WORK,CARRY],name:'max',setup:'worker'});` 

Some notes: 
* you need to define the exact body parts yourself
* setup defines the behaviour your new creep should utilize

**Comment by Krazyfuq** [comment]
@cyberblast thanks a lot i really appreciate any and all help
